igvc_perception/src/pointcloud_filter/fast_segment_filter/fast_segment_filter.cpp

```diff
@@ -0,0 +1,317 @@
+#include <pointcloud_filter/fast_segment_filter/fast_segment_filter.h>
+
+namespace pointcloud_filter
+{
+void FastSegmentFilter::filter(pointcloud_filter::Bundle &bundle)
+{
+  segments_.clear();
+  ground_points_.clear();
+  nonground_points_.clear();
+
+  for (const auto &point : bundle.pointcloud->points)
+  {
+    int segment_id = getAngleFromPoint(point) * config_.num_segments / 360;
+    segments_[segment_id].raw_points_.emplace_back(point);
+  }
+  processSegments();
+  getLinesFromSegments();
+  if (config_.debug_viz)
+  {
+    debugViz();
+  }
+  ground_points_.header.frame_id = "/lidar";
+  nonground_points_.header.frame_id = "/lidar";
+  classifyPoints(ground_points_, nonground_points_);
+  ground_pub_.publish(ground_points_);
+  nonground_pub_.publish(nonground_points_);
+
+  bundle.free_pointcloud->points = std::move(ground_points_.points);
```

Comment:
One more thing: This should be documented more clearly, but `bundle.free_pointcloud` doesn't refer to ground, but rather refers to the angles at which there isn't an occupied obstacle, ie the yellow points here. The `raycast_filter_.filter(bundle);` handles this already, so you don't need to worry about it.
![barrel_covering](https://user-images.githubusercontent.com/4125463/69489840-b5e1d100-0e4c-11ea-8e4a-e3b8cab8bb3e.png)